apps/agreement/contracts/Agreement.sol

```diff
@@ -0,0 +1,872 @@
+        (Action storage action, Setting storage setting) = _getActionAndSetting(_actionId);
+        return _canDispute(action, setting);
+    }
+
+    function canClaimSettlement(uint256 _actionId) external view returns (bool) {
+        (Action storage action, Setting storage setting) = _getActionAndSetting(_actionId);
+        return _canClaimSettlement(action, setting);
+    }
+
+    function canRuleDispute(uint256 _actionId) external view returns (bool) {
+        Action storage action = _getAction(_actionId);
+        return _canRuleDispute(action);
+    }
+
+    function canSubmitEvidence(uint256 _actionId) external view returns (bool) {
+        Action storage action = _getAction(_actionId);
+        return _canSubmitEvidence(action);
+    }
+
+    function canExecute(uint256 _actionId) external view returns (bool) {
+        (Action storage action, Setting storage setting) = _getActionAndSetting(_actionId);
+        return _canExecute(action, setting);
+    }
+
+    function _createChallenge(Action storage _action, address _challenger, uint256 _settlementOffer, bytes _context, Setting storage _setting)
+        internal
+    {
+        // Store challenge
+        Challenge storage challenge = _action.challenge;
+        challenge.challenger = _challenger;
+        challenge.context = _context;
+        challenge.settlementOffer = _settlementOffer;
+        challenge.createdAt = getTimestamp64();
+
+        // Transfer challenge collateral
+        uint256 challengeStake = _getChallengeStake(_setting);
+        require(collateralToken.safeTransferFrom(_challenger, address(this), challengeStake), ERROR_COLLATERAL_TOKEN_TRANSFER_FAILED);
+
+        // Transfer half of the Arbitrator fees
+        (, ERC20 feeToken, uint256 feeAmount) = _setting.arbitrator.getDisputeFees();
+        uint256 arbitratorFees = feeAmount.div(2);
+        challenge.arbitratorFeeToken = feeToken;
+        challenge.arbitratorFeeAmount = arbitratorFees;
+        require(feeToken.safeTransferFrom(_challenger, address(this), arbitratorFees), ERROR_ARBITRATOR_FEE_TRANSFER_FAILED);
+    }
+
+    function _createDispute(Action storage _action, Setting storage _setting) internal returns (uint256) {
+        // Compute missing fees for dispute
+        Challenge storage challenge = _action.challenge;
+        ERC20 challengerFeeToken = challenge.arbitratorFeeToken;
+        uint256 challengerFeeAmount = challenge.arbitratorFeeAmount;
+        (address recipient, ERC20 feeToken, uint256 missingFees, uint256 totalFees) = _getMissingArbitratorFees(
+            _setting,
+            challengerFeeToken,
+            challengerFeeAmount
+        );
+
+        // Create dispute
+        address submitter = _action.submitter;
+        require(feeToken.safeTransferFrom(submitter, address(this), missingFees), ERROR_ARBITRATOR_FEE_DEPOSIT_FAILED);
+        require(feeToken.safeApprove(recipient, totalFees), ERROR_ARBITRATOR_FEE_APPROVAL_FAILED);
```

Comment:
Should we always set the approval to 0 first to prevent approvals from failing in case that the arbitrator didn't transfer all the tokens that it could have? I don't think this should ever happen with Aragon Court, but if it were to happen with an arbitrable it could block all new disputes